### PR TITLE
Feat/fees refund

### LINKED ISSUE #104: pallet-idn-manager: account for fees and deposit diff on `update_subscription`
release or hold the difference in fees and deposit on `update_subscription`
if new fees are lower than already taken fees the tx fails

### LINKED ISSUE #107: pallet-idn-manager: release storage and refund any unused credit when subscription ends
- [ ] on kill_subscription
- [ ] on natural ending

### PR DESCRIPTION
**Pull request by juangirini**
# Feat/fees refund

closes #104 
closes #107 

This pull request implements the fees refund feature and introduces several enhancements and fixes to the subscription management system. The key changes are as follows:

## Summary
- Implements fee refund calculations and handling.
- Enhances subscription management with new events and improved error handling.
- Refactors fee calculation logic to support tiered and linear calculations.
- Adds comprehensive tests for new features and edge cases.

### New Features and Enhancements

1. **Fees Management Improvements**:
   - Introduced `BalanceDirection` and `DiffBalance` structs to represent and handle balance movements.
   - Updated the `FeesManager` trait to include methods for calculating fee differences (`calculate_diff_fees`) and collecting fees with references (`collect_fees`).

2. **Subscription Management Enhancements**:
   - Updated subscription creation, updating, and removal processes to handle fee and deposit differences.
   - Added methods to manage and calculate storage deposits (`calculate_diff_deposit`).
   - Improved the event system to emit detailed subscription-related events such as `FeesCollected` and `SubscriptionRemoved`.

3. **Fee Calculation Logic**:
   - Implemented tiered fee calculation with predefined discount tiers.
   - Enhanced linear fee calculator to handle fee differences.

4. **Testing Enhancements**:
   - Added comprehensive tests for fee calculations, fee collection, subscription updates, and edge cases like insufficient balance.
   - Improved mock setup to better simulate real scenarios.

5. **Code Refactoring**:
   - Refactored existing code for better readability and maintainability.
   - Consolidated common logic and improved error handling.

### CODE REVIEW

**Review by driemworks — APPROVED** [review]
LGTM

File: pallets/idn-manager/src/impls.rs [hunk 1]
```diff
@@ -46,81 +48,135 @@ impl SubscriptionTrait<()> for () {
...
+	for FeesManagerImpl<T, B, S, Balances>
 where
-	Balances::Balance: From<BlockNumber>,
 	Balances::Reason: From<HoldReason>,
+	Balances::Balance: From<u64>,
 {
-	fn calculate_subscription_fees(amount: BlockNumber) -> Balances::Balance {
-		let base_fee = B::get();
-		base_fee.saturating_mul(amount.into())
+	fn calculate_subscription_fees(credits: &u64) -> Balances::Balance {
```
Comment (carloskiron) [inline pallets/idn-manager/src/impls.rs:68]: @juangirini I suggest to add comments to these functions involving calculations to get the methodology in the same context than the code and make a better sense of it.t

**Review by carloskiron — APPROVED** [review]
@juangirini looks good to me. Coming along nicely, only left a comment related to a nice to have comments on the implementation of fees calculation and associated functions.

File: pallets/idn-manager/src/impls.rs [hunk 2]
```diff
@@ -46,81 +48,135 @@ impl SubscriptionTrait<()> for () {
...
+
+		const BASE_FEE: u64 = 100;
+
+		let mut total_fee = 0u64;
+		let mut remaining_credits = *credits;
+
+		for (i, &(current_tier_start, current_tier_discount)) in TIERS.iter().enumerate() {
+			// If no remaining credits or the tier starts above the requested credits, exit loop.
+			if remaining_credits == 0 || credits < &current_tier_start {
+				break;
```
Comment (colemanirby) [inline pallets/idn-manager/src/impls.rs:86]: should we differentiate between these two scenarios?

**Review by colemanirby — COMMENTED** [review]
LGTM

File: pallets/idn-manager/src/impls.rs (hunk 1 above)
Comment (juangirini) [reply pallets/idn-manager/src/impls.rs:98]: I have improved the docs and added some doctests that don't compile. I would try and make them compile as part of the documentation issue https://github.com/ideal-lab5/idn-sdk/issues/78
